provider/godaddy/client.go

```diff
@@ -0,0 +1,300 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package godaddy
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"sigs.k8s.io/external-dns/pkg/apis/externaldns"
+)
+
+// DefaultTimeout api requests after 180s
+const DefaultTimeout = 180 * time.Second
+
+// Errors
+var (
+	ErrAPIDown = errors.New("godaddy: the GoDaddy API is down")
+)
+
+// APIError error
+type APIError struct {
+	Code    string
+	Message string
+}
+
+func (err *APIError) Error() string {
+	return fmt.Sprintf("Error %s: %q", err.Code, err.Message)
+}
+
+// Logger is the interface that should be implemented for loggers that wish to
+// log HTTP requests and HTTP responses.
+type Logger interface {
+	// LogRequest logs an HTTP request.
+	LogRequest(*http.Request)
+
+	// LogResponse logs an HTTP response.
+	LogResponse(*http.Response)
+}
+
+// Client represents a client to call the GoDaddy API
+type Client struct {
+	// APIKey holds the Application key
+	APIKey string
+
+	// APISecret holds the Application secret key
+	APISecret string
+
+	// API endpoint
+	APIEndPoint string
+
+	// Client is the underlying HTTP client used to run the requests. It may be overloaded but a default one is instanciated in ``NewClient`` by default.
+	Client *http.Client
+
+	// Logger is used to log HTTP requests and responses.
+	Logger Logger
+
+	Timeout time.Duration
+}
+
+// NewClient represents a new client to call the API
+func NewClient(useOTE bool, apiKey, apiSecret string) (*Client, error) {
+	var endpoint string
+
+	if useOTE {
+		endpoint = " https://api.ote-godaddy.com"
+	} else {
+		endpoint = "https://api.godaddy.com"
+	}
+
+	client := Client{
+		APIKey:      apiKey,
+		APISecret:   apiSecret,
+		APIEndPoint: endpoint,
+		Client:      &http.Client{},
+		Timeout:     DefaultTimeout,
+	}
+
+	// Get and check the configuration
+	if err := client.validate(); err != nil {
+		return nil, err
+	}
+	return &client, nil
+}
+
+//
+// Common request wrappers
+//
+
+// Get is a wrapper for the GET method
+func (c *Client) Get(url string, resType interface{}) error {
```

Comment:
godaddy doesn't provide any sdk for handling domains and records? I'm suprised, this might be a bit hard to maintain.

Reply:
GoDaddy doesn't provide any sdk for any language.

I found some some custom implementation in golang such as this one: https://github.com/kryptoslogic/godaddy-domainclient but any provided the missing delete DNS record API call.

In fact to delete a record, the only method is to replace all entries in the zone with all zone records minus the deleted record.